client/ext/bigquery/current-timestamp-optimizer.js

```diff
@@ -0,0 +1,186 @@
+
+  /**
+   * Constants describing the types of granularity supported on timestamps.
+   * @enum
+   */
+  bigquery.CurrentTimestampGranularity = {
+    YEAR: 'YEAR',
+    MONTH: 'MONTH',
+    WEEK: 'WEEK',
+    DAY: 'DAY',
+    HOUR: 'HOUR'
+  };
+  const CurrentTimestampGranularity = bigquery.CurrentTimestampGranularity;
+
+  /**
+   * Describes the properties for configuring a CurrentTimestampOptimizer.
+   *
+   * @constructor
+   */
+  bigquery.CurrentTimestampOptimizerConfigModel = class {
+    constructor(enabled=null, granularity=null) {
+      /**
+       * If true, the optimizer is enabled.  If false, it will not be applied.
+       *
+       * @export {bool}
+       */
+      this.enabled = enabled;
+
+      /**
+       * Defines the granularity of the optimizer.  See the module docstring for more.
+       * 
+       * @export {!CurrentTimestampGranularity}
+       */
+      this.granularity = granularity;
+    }
+  }
+
+  /**
+   * See the module docstring for a description of this optimizer.
+   *
+   * @constructor
+   */
+  bigquery.CurrentTimestampOptimizer = class {
+    constructor() {
+      
+    }
+    
+    /**
+     * Returns true if the optimizer should be applied to the widget, otherwise false.
+     *
+     * @param {!DashboardModel} dashboard The active dashboard.
+     * @param {!WidgetModel} widget The widget to apply query optimizations against.
+     * @return {boolean} True if the optimizer should be applied to the current widget,
+     *     otherwise false.
+     */
+    canApply(dashboard, widget) {
+      if (goog.isDefAndNotNull(widget.datasource.config.bigQuery.optimizeCurrentTimestamp.enabled)) {
+        return widget.datasource.config.bigQuery.optimizeCurrentTimestamp.enabled;
+      }
+      
+      return dashboard.config.bigQuery.optimizeCurrentTimestamp.enabled;
+    }
+
+    /**
+     * Returns the effective granularity for a widget on a dashboard.  See the class docstring
+     * for more information on how granularity affects date calculation.
+     *
+     * If the widget .enabled setting is true and the widget's .granularity is specified, then
+     * the widget's .granularity is used.  Otherwise, the dashboard's .granularity is used.
+     * 
+     * @param {!DashboardModel} dashboard The active dashboard.
+     * @param {!WidgetModel} widget The widget to apply query optimizations against.
+     * @return {!CurrentTimestampGranularity} The granularity to apply to the widget's
+     *     CURRENT_TIMESTAMP() functions.
+     */
+    getEffectiveGranularity(dashboard, widget) {
+      let widgetSetting = widget.datasource.config.bigQuery.optimizeCurrentTimestamp;
+      let dashboardSetting = dashboard.config.bigQuery.optimizeCurrentTimestamp;
+
+      if (goog.isDefAndNotNull(widgetSetting.granularity)) {
+        return widgetSetting.granularity;
+      } else {
+        return dashboardSetting.granularity;
+      }
+    }
+
+    /**
+     * Returns the current date.  Used for mocking purposes.
```

Comment:
Mocking out the Date constructor was problematic because you can't mock out the timezone as well, so you never actually know what "real" time you're going to get.  Originally attempted this solution without using a date provider for explicit mocking, and it didn't work out.

As to what is cleaner, I tend to side with minimizing the usage complexity (which passing in Date would do) rather than minimizing the implementation complexity.